server/sonar-server/src/test/java/org/sonar/server/email/ws/UpdateActionTest.java

```diff
@@ -0,0 +1,173 @@
+import static org.sonar.db.property.PropertyTesting.newGlobalPropertyDto;
+
+public class UpdateActionTest {
+
+  @Rule
+  public ExpectedException expectedException = ExpectedException.none();
+
+  @Rule
+  public UserSessionRule userSession = UserSessionRule.standalone();
+
+  @Rule
+  public DbTester db = DbTester.create(System2.INSTANCE);
+
+  DbClient dbClient = db.getDbClient();
+  PropertyDbTester propertyDb = new PropertyDbTester(db);
+  org.sonar.api.config.Settings settings = new org.sonar.api.config.Settings();
+
+  WsActionTester ws = new WsActionTester(new UpdateAction(dbClient, userSession, settings));
+
+  @Test
+  public void update_email_settings() throws Exception {
+    setUserAsSystemAdmin();
+
+    executeRequest("smtp.gmail.com", 25, "starttls", "john", "12345", "noreply@email.com", "[EMAIL]");
+
+    assertSetting("email.smtp_host.secured", "smtp.gmail.com");
+    assertSetting("email.smtp_port.secured", "25");
+    assertSetting("email.smtp_secure_connection.secured", "starttls");
+    assertSetting("email.smtp_username.secured", "john");
+    assertSetting("email.smtp_password.secured", "12345");
+    assertSetting("email.from", "noreply@email.com");
+    assertSetting("email.prefix", "[EMAIL]");
+  }
+
+  @Test
+  public void does_not_save_settings_when_no_value_sent() throws Exception {
+    setUserAsSystemAdmin();
+
+    executeRequest(null, null, null, null, null, null, null);
+
+    assertThat(db.countRowsOfTable(db.getSession(), "properties")).isZero();
+  }
+
+  @Test
+  public void remove_existing_settings_when_no_value_sent() throws Exception {
+    setUserAsSystemAdmin();
+    addSetting("email.smtp_host.secured", "smtp.gmail.com");
+    addSetting("email.smtp_port.secured", "25");
+    addSetting("email.smtp_secure_connection.secured", "starttls");
+    addSetting("email.smtp_username.secured", "john");
+    addSetting("email.smtp_password.secured", "12345");
+    addSetting("email.from", "noreply@email.com");
+    addSetting("email.prefix", "[EMAIL]");
+
+    executeRequest(null, null, null, null, null, null, null);
+
+    assertThat(db.countRowsOfTable(db.getSession(), "properties")).isZero();
+  }
+
+  @Test
+  public void fail_when_secure_param_is_invalid() {
+    setUserAsSystemAdmin();
+
+    expectedException.expect(IllegalArgumentException.class);
+
+    executeRequest(null, null, "unknown", null, null, null, null);
+  }
+
+  @Test
+  public void fail_when_insufficient_privileges() {
+    userSession.anonymous().setGlobalPermissions(GlobalPermissions.QUALITY_PROFILE_ADMIN);
+    expectedException.expect(ForbiddenException.class);
+
+    ws.newRequest().execute();
+  }
+
+  @Test
+  public void test_ws_definition() {
+    WebService.Action action = ws.getDef();
+    assertThat(action).isNotNull();
+    assertThat(action.isInternal()).isTrue();
+    assertThat(action.isPost()).isTrue();
+    assertThat(action.responseExampleAsString()).isNull();
+    assertThat(action.params()).hasSize(7);
```

Comment:
do we want to test the parameters key? do we want to test the possible values of ssl? It may be overkill